morphman/automated_landmarking/automated_landmarking_bogunovic.py

```diff
@@ -0,0 +1,175 @@
+    """
+
+    if resampling_step is not None:
+        centerline = vmtk_resample_centerline(centerline, length=resampling_step)
+
+    if approximation_method == "vmtk":
+        line = vmtk_compute_centerline_attributes(centerline)
+        line = vmtk_compute_geometric_features(line, smooth_line, factor=smoothing_factor, iterations=iterations)
+        curvature = get_point_data_array("Curvature", line)
+        curvature = gaussian_filter(curvature, 2)
+
+    elif approximation_method == "disc":
+        neigh = 20
+        line = vmtk_compute_centerline_attributes(centerline)
+        line = vmtk_compute_geometric_features(line, smooth_line, factor=smoothing_factor, iterations=iterations)
+        line, curvature__ = compute_discrete_derivatives(line, neigh=neigh)
+        curvature = []
+        for c in curvature__:
+            curvature.append([c])
+        curvature = np.array(curvature)
+
+    elif approximation_method == "spline":
+        line, max_point_ids, min_point_ids = spline_centerline_and_compute_geometric_features(centerline, smooth_line,
+                                                                                              nknots)
+        curvature = get_point_data_array("Curvature", line)
+
+    if approximation_method != "spline":
+        max_point_ids = list(argrelextrema(curvature, np.greater)[0])
+        min_point_ids = list(argrelextrema(curvature, np.less)[0])
+        get_k1k2_basis(curvature, line)
+
+    length = get_curvilinear_coordinate(line)
+    k1 = get_point_data_array("k1", line)
+    k2 = get_point_data_array("k2", line)
+
+    # Remove a min / max point that is in reality a saddle point
+    for i in min_point_ids:
+        for j in max_point_ids:
+            if abs(i - j) < 5 and abs(curvature[i] - curvature[j]) < 0.01:
+                if i in min_point_ids:
+                    min_point_ids.remove(i)
+                if j in max_point_ids:
+                    max_point_ids.remove(j)
+
+    k1_points = k1[max_point_ids]
+    k2_points = k2[max_point_ids]
+    k_points = np.zeros((k1_points.shape[0], 2))
+    k_points[:, 0] = k1_points[:, 0]
+    k_points[:, 1] = k2_points[:, 0]
+    theta = np.zeros(k1_points.shape[0] - 1)
+    for i in range(theta.shape[0]):
+        a = k_points[i, :] / np.sqrt(np.sum(k_points[i, :] ** 2))
+        b = k_points[i + 1, :] / np.sqrt(np.sum(k_points[i + 1, :] ** 2))
+        theta[i] = math.acos(np.dot(a, b))
+        theta[i] = theta[i] * 180 / math.pi
+
+    # Tolerance parameters from Bogunovic et al. (2012)
+    tol_ant_post = 60
+    tol_sup_ant = 45
+    tol_post_inf = 45
+    tol_inf_end = 110
+
+    # Find coronal coordinate and maximum (within anterior bend)
+    coordinates = get_centerline_coordinates(line, length)
+    coronal_coordinate = coordinates[coronal_axis]
+    max_coronal_coordinate_id = get_maximum_coronal_coordinate(coronal_coordinate, length)
+
+    def find_interface(start, direction, tol, part):
```

Comment:
Does this have to be a nested function? Would it not be better to move it outside and give it a docstring?

Reply:
Indeed. Separate both `find_interface` methods outside + docstring 👍